multicore_runtime/dispatcher.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""A WSGI app that, once configured, dispatches requests to user apps."""
+
+import logging
+import re
+
+from werkzeug.wrappers import Request
+from werkzeug.wrappers import Response
+
+
+def dispatcher(handlers):
+  """Accepts handlers and returns a WSGI app that dispatches requests to them.
+
+  Args:
+    handlers: a list of handlers as produced by
+      wsgi_utils.load_user_scripts_into_handlers: a list of tuples of
+      (url, script, app).
+
+  Returns:
+    A WSGI app that dispatches to the user apps specified in the input.
+  """
+
+  def dispatch(wsgi_env, start_response):
+    """Handle one request."""
+    request = Request(wsgi_env)
+    for url, script, app in handlers:  # pylint: disable=unused-variable
+      matcher = re.match(url, request.path)
+      if matcher and matcher.end() == len(request.path):
+        if app is not None:
```

Comment:
Also thinking you could simplify this a bit, since in both cases you return a callable with the same two arguments:

```python
  if app:
    response = app
  else:
    response = Response(...)
  return response(wsgi_env, start_response)
```

Or, if you don't mind overwriting `app`:

```python
  if not app:
    app = Response(...)
  return app(wsgi_env, start_response)
```

Though what you currently have is certainly fine as is, this is just a suggestion.